keps/sig-cli/3104-introduce-kuberc/README.md

```diff
@@ -0,0 +1,896 @@
+nitty-gritty.
+-->
+
+We propose introducing a new file that will be versioned and intended for user provided preferences. This new file is entirely opt-in.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user I would like to specify preferences per client (device) without my kubeconfig overriding them when I access different clusters.
+
+#### Story 2
+
+As a user I would like to opt in to enforcing delete confirmation or colored output on my local client but not disrupt my CI pipeline with unexpected prompts or output.
+
+https://groups.google.com/g/kubernetes-dev/c/y4Q20V3dyOk
+
+https://github.com/kubernetes/kubectl/issues/524
+
+#### Story 3
+
+[UNRESOLVED] As a user I would like to use different preferences per context.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+| Risk | Impact | Mitigation |
+| --- | --- | --- |
+| Confusing users with a new config file | Low | Documentation and education |
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+During alpha this feature will be enabled through the `ENABLE_KUBERC=true` environment variable. The file will default to being located in `~/.kube/kuberc`. A flag will allow overriding this default location with a path i.e. `kubectl --kuberc /var/kube/rc`.
+
+Three initial top level keys are proposed.
+
+* `apiVersion` to determine the version of the config.
+* `kind` to keep consistency with Kubernetes resources.
+* `command`
+  * `aliases` for users to declare their own aliases for commands with flags and values.
+  * `overrides` for users to set default flags to apply to commands.
+
+`command.aliases` will not be allowed to override builtins but take precedence of plugins i.e. builtins -> aliases -> plugins. Additional flags and values will be appended to the end of the aliased command. It is the responsibility of the user to define aliases with this in mind.
```

Comment:
Could we consider different aliasing rules/types? For example aliasing namespaces:

instead of `oc get pods -n kube-system`, we would do `oc get pods -n ks`

I suppose we could also move defining such aliases to flag overrides instead of aliases section. Although I worry generic aliases for flags would be too much in the beginning and could cause problems when running scripts, etc.

Reply:
I don't want to go that far, I'd start simple with just command aliases.

Reply:
We can always consider other bits as we move forward.